File: Data Structures & Algorithms/find-peak-element/submission-1.java
class Solution {
    public int findPeakElement(int[] nums) {
        int l = 0 , r = nums.length -1 , mid = r /2;
        while(l < r){
            if(mid + 1 < nums.length && mid - 1 >= 0){
                if(nums[mid-1] < nums[mid] && nums[mid+1] < nums[mid]){
                    return mid;
                }
                else if(nums[mid+1] > nums[mid-1]){
                    l = mid +1;
                }   
                else{
                    r = mid -1;
                }
            }
            else if(mid + 1 < nums.length && nums[mid] > nums[mid+1] || mid - 1 >= 0 && nums[mid] > nums[mid-1]){
                return mid;
            }
            else{
                if(mid + 1 < nums.length){
                    l = mid +1;
                }
                else{
                    r = mid -1;
                }
            }
            mid = (r - l) / 2 + l;
        }
        System.out.print("out of binary!");
        return mid;
    }
}

File: Data Structures & Algorithms/permutations/submission-1.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void backtrack(vector<int> arr , vector<int> per){
        if(arr.size() == 1){
            per.push_back(arr[0]);
            ans.push_back(per);
            return;
        }

        for(int i = 0 ; i<arr.size();i++){
            /* cout << "adding:"<<arr[i]<<endl; */
            int temp = arr[i];
            arr.erase(arr.begin() + i);
            per.push_back(temp);
            backtrack(arr , per);
            /* cout << "Sending: {";
            for(auto x:arr){
                cout << x << ",";
            }
            cout<<"}\n"; */

            arr.insert(arr.begin() + i , temp);
            per.pop_back();
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        if(!nums.size()){
            return {{}};
        }
        if(nums.size() == 1){
            return {{nums[0]}};
        }

        vector<int> arr(nums);
        vector<int> per;
        for(int i =0 ; i < arr.size() ;i++){
            /* cout <<"i:"<<i << " adding:"<<arr[i]<<endl; */
            int temp = arr[i];
            arr.erase(arr.begin() + i);
            per.push_back(temp);
            backtrack(arr , per);
            /* cout << "Sending: {";
            for(auto x:arr){
                cout << x << ",";
            }
            cout<<"}\n"; */

            arr.insert(arr.begin() + i , temp);
            per.pop_back();
        }

        return ans;
    }
};
